NIFI-4708 Add Registry support to encrypt-config

kevdoran opened the pull request:
Adds support for NiFI Registry config files to the encrypt-config tool in NiFi Toolkit.

Also adds decryption capability to encrypt-config tool.


Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

kevdoran commented:
FYI @alopresto and @bbende - This is ready to be reviewed for merge to master.

There are still a few test cases I would like to add, time permitting, for xml files, but that could be addressed in a follow up PR, along with additional refinement and a refactoring that @alopresto and I have discussed.

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy:
```diff
@@ -0,0 +1,322 @@
+
+        } catch (Exception e) {
+            logger.error("Encountered an error: ${e.getMessage()}")
+            logger.debug("", e) // stack trace only when verbose enabled
```
I understand why this was done but I think the better logic is:

```
if (isVerboseEnabled()) {
    logger.error("", e)
}
```
Some people/tools are set up to extract messages based on levels, and this is semantically an error.

kevdoran replied:
Good call.
- [ ] Change error logging to got to error, wrap in conditional if it depends on verbose

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy:
```diff
@@ -0,0 +1,322 @@
+            throw new RuntimeException("Failed to load and decrypt input file.")
+        }
+
+        if (config.outputToFile) {
```
I don't think `decrypt` should allow output to a file. At this stage, it is safer to require the output goes to STDOUT to avoid decrypting files in place.

kevdoran replied:
Now that I think about it, I agree with you. That would be the better/safer approach for this initial offering of the decrypt capability. Will update

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy:
```diff
@@ -0,0 +1,322 @@
+
+    }
+
+    static class DecryptConfiguration {
```
Should this implement the `Configuration` interface?

kevdoran replied:
Yes.
- [ ] Adopt the Configuration interface where applicable to convey intended class structure to future developers as the Configuration interface is filled in more.

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy:
```diff
@@ -0,0 +1,322 @@
+                fileType = FileType.valueOf(rawOptions.t)
+            }
+
+            if (rawOptions.o) {
```
As per discussion, this option can be removed.

kevdoran replied:
Yep, as explained above, will remove the -R option from `NiFiRegistryDecryptMode`

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy:
```diff
@@ -0,0 +1,322 @@
+        private void determineInputFileFromRemainingArgs() {
+            String[] remainingArgs = this.rawOptions.getInner().getArgs()
+            if (remainingArgs.length == 0) {
+                throw new RuntimeException("Missing argument: Input file must provided.")
```
Input file must *be* provided.

kevdoran replied:
- [ ] fix typo

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/DecryptMode.groovy:
```diff
@@ -0,0 +1,322 @@
+            if (remainingArgs.length == 0) {
+                throw new RuntimeException("Missing argument: Input file must provided.")
+            } else if (remainingArgs.length > 1) {
+                throw new RuntimeException("Too many arguments: Please specify exactly one input file in addition to the options.")
```
Is order important here? If `$ encrypt-config.sh --decrypt -b bootstrap.conf file_to_decrypt.properties` works but `$ encrypt-config.sh file_to_decrypt.properties --decrypt -b bootstrap.conf` doesn't, I think this message should indicate that.

kevdoran replied:
This is actually correct if you call `DecryptMode.run(String[])` (not exposed to end users in this commit). The usage you're thinking of is through `NiFiRegistryDecryptMode.run(String[])`, which does its own argument parsing and calls `DecryptMode.run(DecryptConfiguration)`

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/EncryptConfigLogger.groovy:
```diff
@@ -0,0 +1,93 @@
+
+        return defaultProperties
+    }
+
```
Maybe just run a "Format Code" command on this one to remove extra whitespace.

kevdoran replied:
- [ ] Format code

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/EncryptConfigMain.groovy:
```diff
@@ -0,0 +1,145 @@
+                false)
+        System.out.println()
+
+//        String footer = """
```
Remove dead code.

kevdoran replied:
- [ ] remove dead code in EMC

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/EncryptConfigMain.groovy:
```diff
@@ -0,0 +1,145 @@
+                printUsageAndExit(EXIT_STATUS_FAILURE)
+            }
+        } catch (Throwable t) {
+            logger.debug("", t)
```
Change this to `logger.error()` as well.

kevdoran replied:
- [ ] use `logger.error()` in ECM

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryDecryptMode.groovy:
```diff
@@ -0,0 +1,143 @@
+            config.inputFilePath = options.r
+            config.fileType = FileType.properties  // disables auto-detection, which is still experimental
+
+            // one of [--oldPassword, --oldKey] or [-p, -k, -b <file]
```
Typo in comment -- `[-p, -k, -b <file>]`.

kevdoran replied:
- [ ] fix typo

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryDecryptMode.groovy:
```diff
@@ -0,0 +1,143 @@
+            config.fileType = FileType.properties  // disables auto-detection, which is still experimental
+
+            // one of [--oldPassword, --oldKey] or [-p, -k, -b <file]
+            String keyHex = null
```
I think this logic should be extracted to a method and condensed. Something like this (note: should `oldPassword` and `oldKey` be accepted for decryption?):

```
Index: nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryDecryptMode.groovy
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryDecryptMode.groovy	(date 1515148235000)
+++ nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryDecryptMode.groovy	(revision )
@@ -67,77 +67,112 @@
             config.inputFilePath = options.r
             config.fileType = FileType.properties  // disables auto-detection, which is still experimental
 
-            // one of [--oldPassword, --oldKey] or [-p, -k, -b <file]
-            String keyHex = null
-            String password = null
-            config.keySource = null
-            if (options.oldPassword) {
-                if (config.keySource != null) {
+
+            config.key = retrieveKey(config, options)
+
+            config.decryptionProvider = new AESSensitivePropertyProvider(config.key)
+
+            /* Optional fields when using --decrypt */
+            // -R outputRegistryPropertiesFile (-R)
+            if (options.R) {
+                config.outputToFile = true
+                config.outputFilePath = options.R
+            }
+
+            run(config)
+
+        } catch (Exception e) {
+            logger.error("Encountered an error: ${e.getMessage()}")
+            logger.debug("", e)  // only print stack trace when verbose is enabled
+            EncryptConfigMain.printUsageAndExit(EncryptConfigMain.EXIT_STATUS_FAILURE)
+        }
+    }
+
+    private String retrieveKey(DecryptConfiguration config, OptionAccessor options) {
+        // one of [--oldPassword, --oldKey] or [-p, -k, -b <file]
+        String keyHex = null
+        String password = null
+        config.keySource = null
+
+        Map<String, Configuration.KeySource> keySourceMap = [
+                oldPassword: Configuration.KeySource.PASSWORD,
+                oldKey     : Configuration.KeySource.KEY_HEX,
+                p          : Configuration.KeySource.PASSWORD,
+                k          : Configuration.KeySource.KEY_HEX,
+//                b          : Configuration.KeySource.BOOTSTRAP_FILE
+        ]
+
+        keySourceMap.each { String name, Configuration.KeySource ks ->
+            if (options."$name") {
+                if (config.keySource) {
                     throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
                 }
-                config.keySource = Configuration.KeySource.PASSWORD
-                keyHex = options.getInner().getOptionValue("oldPassword")
-            }
-            if (options.oldKey) {
-                if (config.keySource != null) {
-                    throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
+                config.keySource = ks
+                if (ks == Configuration.KeySource.PASSWORD) {
+                    password = options.getInner().getOptionValue(name)
+                } else {
+                    keyHex = options.getInner().getOptionValue(name)
                 }
-                config.keySource = Configuration.KeySource.KEY_HEX
-                keyHex = options.getInner().getOptionValue("oldKey")
             }
-            if (options.p) {
-                if (config.keySource != null) {
-                    throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
-                }
-                config.keySource = Configuration.KeySource.PASSWORD
-                password = options.getInner().getOptionValue("p")
-            }
-            if (options.k) {
-                if (config.keySource != null) {
-                    throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
-                }
-                config.keySource = Configuration.KeySource.KEY_HEX
-                keyHex = options.getInner().getOptionValue("k")
-            }
+        }
 
-            if (config.keySource) {
-                config.key = ToolUtilities.determineKey(keyHex, password, Configuration.KeySource.PASSWORD.equals(config.keySource))
-            }
+        // Convert the password to a key (if necessary) and store in the config object
+        if (config.keySource) {
+            config.key = ToolUtilities.determineKey(keyHex, password, config.keySource == Configuration.KeySource.PASSWORD)
+        }
+
+//        if (options.oldPassword) {
+//            if (config.keySource != null) {
+//                throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
+//            }
+//            config.keySource = Configuration.KeySource.PASSWORD
+//            keyHex = options.getInner().getOptionValue("oldPassword")
+//        }
+//        if (options.oldKey) {
+//            if (config.keySource != null) {
+//                throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
+//            }
+//            config.keySource = Configuration.KeySource.KEY_HEX
+//            keyHex = options.getInner().getOptionValue("oldKey")
+//        }
+//        if (options.p) {
+//            if (config.keySource != null) {
+//                throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
+//            }
+//            config.keySource = Configuration.KeySource.PASSWORD
+//            password = options.getInner().getOptionValue("p")
+//        }
+//        if (options.k) {
+//            if (config.keySource != null) {
+//                throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
+//            }
+//            config.keySource = Configuration.KeySource.KEY_HEX
+//            keyHex = options.getInner().getOptionValue("k")
+//        }
+
+//        if (config.keySource) {
+//            config.key = ToolUtilities.determineKey(keyHex, password, Configuration.KeySource.PASSWORD.equals(config.keySource))
+//        }
 
-            if (options.b) {
-                if (config.keySource != null) {
-                    throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
-                }
-                config.keySource = Configuration.KeySource.BOOTSTRAP_FILE
-                config.inputBootstrapPath = options.b
+        if (options.b) {
+            if (config.keySource) {
+                throw new RuntimeException("Invalid options: Only one of [--oldPassword, --oldKey, -b, -p, -k] is allowed for specifying the decryption password/key.")
+            }
+            config.keySource = Configuration.KeySource.BOOTSTRAP_FILE
+            config.inputBootstrapPath = options.b
 
-                logger.debug("Checking expected NiFi Registry bootstrap.conf format")
-                config.key = BootstrapUtil.extractKeyFromBootstrapFile(config.inputBootstrapPath, BootstrapUtil.REGISTRY_BOOTSTRAP_KEY_PROPERTY)
+            logger.debug("Checking expected NiFi Registry bootstrap.conf format")
+            config.key = BootstrapUtil.extractKeyFromBootstrapFile(config.inputBootstrapPath, BootstrapUtil.REGISTRY_BOOTSTRAP_KEY_PROPERTY)
 
-                // check we have found the key
-                if (config.key) {
-                    logger.debug("Master key found in ${config.inputBootstrapPath}. This key will be used for decryption operations.")
-                } else {
-                    logger.warn("Bootstrap Conf flag present, but master key could not be found in ${config.inputBootstrapPath}.")
-                }
-            }
+            // check we have found the key
+            if (config.key) {
+                logger.debug("Master key found in ${config.inputBootstrapPath}. This key will be used for decryption operations.")
+            } else {
+                logger.warn("Bootstrap conf flag present, but master key could not be found in ${config.inputBootstrapPath}.")
+            }
+        }
 
-            config.decryptionProvider = new AESSensitivePropertyProvider(config.key)
-
-            /* Optional fields when using --decrypt */
-            // -R outputRegistryPropertiesFile (-R)
-            if (options.R) {
-                config.outputToFile = true
-                config.outputFilePath = options.R
-            }
-
-            run(config)
-
-        } catch (Exception e) {
-            logger.error("Encountered an error: ${e.getMessage()}")
-            logger.debug("", e)  // only print stack trace when verbose is enabled
-            EncryptConfigMain.printUsageAndExit(EncryptConfigMain.EXIT_STATUS_FAILURE)
-        }
+        config.key
     }
 
 }

```

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryDecryptMode.groovy:
```diff
@@ -0,0 +1,143 @@
+            config.decryptionProvider = new AESSensitivePropertyProvider(config.key)
+
+            /* Optional fields when using --decrypt */
+            // -R outputRegistryPropertiesFile (-R)
```
As per discussion, remove this option.

kevdoran replied:
- [ ] remove -R

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryDecryptMode.groovy:
```diff
@@ -0,0 +1,143 @@
+
+        } catch (Exception e) {
+            logger.error("Encountered an error: ${e.getMessage()}")
+            logger.debug("", e)  // only print stack trace when verbose is enabled
```
Change to `logger.error()`.

kevdoran replied:
- [ ] `logger.error()`

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryMode.groovy:
```diff
@@ -0,0 +1,383 @@
+
+    @Override
+    void run(String[] args) {
+        logger.warn("The NiFi Registry capabilities of this tool is still considered experimental. The results should be manually verified.")
```
...capabilities of this tool *are* still considered experimental.

kevdoran replied:
- [ ] fix typo

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryMode.groovy:
```diff
@@ -0,0 +1,383 @@
+
+        logger.debug("(src)  bootstrap.conf:         ${config.inputBootstrapPath}")
+        logger.debug("(dest) bootstrap.conf:         ${config.outputBootstrapPath}")
+        logger.debug("(src)  nifi.properties:        ${config.inputNiFiRegistryPropertiesPath}")
```
This line and the next should print `nifi-registry.properties`.

kevdoran replied:
- [ ] fix file label

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/NiFiRegistryMode.groovy:
```diff
@@ -0,0 +1,383 @@
+
+            writingKeyToBootstrap = (usingPassword || usingRawKeyHex || rawOptions.B)
+            if (writingKeyToBootstrap) {
+                outputBootstrapPath = rawOptions.B ? rawOptions.B : inputBootstrapPath
```
In the future, the Elvis operator condenses this in Groovy (not necessary to make the change for this PR). 

`outputBootstrapPath = rawOptions.B ?: inputBootstrapPath`

kevdoran replied:
Oh cool, I didn't know about that Groovy idiom - thanks!

- [ ] (optional, time permitting) change to using Elvis operator where applicable

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/util/BootstrapUtil.groovy:
```diff
@@ -0,0 +1,132 @@
+    static String extractKeyFromBootstrapFile(String inputBootstrapPath, String bootstrapKeyPropertyName) throws IOException {
+
+        File inputBootstrapConfFile
+        if (!(inputBootstrapPath && (inputBootstrapConfFile = new File(inputBootstrapPath)).exists() && inputBootstrapConfFile.canRead())) {
```
Not required for this PR, but in the future this is a good anti-pattern for extracting the control logic to a boolean checker method like `isInputBootstrapConfValid()`.

kevdoran replied:
Good catch, will change this to use the utility method `ToolUtilities.canRead(File)` method used elsewhere

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/util/NiFiRegistryAuthorizersXmlEncryptor.groovy:
```diff
@@ -0,0 +1,106 @@
+        super(encryptionProvider, decryptionProvider)
+    }
+
+    @Override
```
Add Javadoc explaining operation of overriding method and why custom implementation is necessary.

alopresto commented on nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/toolkit/encryptconfig/util/NiFiRegistryIdentityProvidersXmlEncryptor.groovy:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    String encrypt(String plainXmlContent) {
```
Add Javadoc explaining operation of overriding method and why custom implementation is necessary.

kevdoran replied:
- [ ] Add Javadoc to sub-classes of XMLEncryptor

kevdoran commented:
@alopresto thanks for the thorough review! I've pushed an update that addresses your comments. It also adds a lot of test cases (see NiFiRegistryModeSpec and NiFiRegistryDecryptModeSpec), which cover all the functionality that I intend to expose in this version (ie, [--nifiRegistry [--decrypt] [options]]. Given that I had time to do more testing, I also removed the "experimental" warning output when the new modes are run. Let me know what you think and if you see anything else to improve in this PR.

alopresto commented:
Ran `contrib-check` and all tests pass. 

Ran a few scenarios using the tool. Everything looks good. The only slight issue was when I intentionally ran with the wrong arguments and it looks like the error message contains the whole option object and stacktrace in addition to the flag, but it's clear enough for a user to understand the issue. 

```
hw12203:...assembly/target/nifi-toolkit-1.5.0-SNAPSHOT-bin/nifi-toolkit-1.5.0-SNAPSHOT (pr2376) alopresto
🔓 271505s @ 11:01:23 $ ./bin/encrypt-config.sh --nifiRegistry -b /Users/alopresto/Workspace/registry/nifi-registry-assembly/target/nifi-registry-0.0.1-SNAPSHOT-bin/nifi-registry-0.0.1-SNAPSHOT/conf/bootstrap.conf -r /Users/alopresto/Workspace/registry/nifi-registry-assembly/target/nifi-registry-0.0.1-SNAPSHOT-bin/nifi-registry-0.0.1-SNAPSHOT/conf/nifi-registry.properties -R /Users/alopresto/Workspace/registry/nifi-registry-assembly/target/nifi-registry-0.0.1-SNAPSHOT-bin/nifi-registry-0.0.1-SNAPSHOT/conf/nifi-registry-encrypted.properties -v --decrypt
2018-01-08 11:01:40 DEBUG EncryptConfigLogger: Verbose mode is enabled (goes to stderr by default).
2018-01-08 11:01:40 ERROR NiFiRegistryDecryptMode: Encountered an error: Invalid options: --decrypt cannot be used with [[ option: R outputNifiRegistryProperties  [ARG] :: The destination nifi-registry.properties file containing protected config values. :: class java.lang.String ]]. It should only be used with -r and one of [-p, -k, -b].
java.lang.RuntimeException: Invalid options: --decrypt cannot be used with [[ option: R outputNifiRegistryProperties  [ARG] :: The destination nifi-registry.properties file containing protected config values. :: class java.lang.String ]]. It should only be used with -r and one of [-p, -k, -b].
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:80)
	at org.codehaus.groovy.reflection.CachedConstructor.doConstructorInvoke(CachedConstructor.java:74)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrap.callConstructor(ConstructorSite.java:84)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:60)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:235)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:247)
	at org.apache.nifi.toolkit.encryptconfig.NiFiRegistryDecryptMode.run(NiFiRegistryDecryptMode.groovy:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSite.invoke(PogoMetaMethodSite.java:169)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:71)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.apache.nifi.toolkit.encryptconfig.EncryptConfigMain.main(EncryptConfigMain.groovy:109)
Invalid options: --decrypt cannot be used with [[ option: R outputNifiRegistryProperties  [ARG] :: The destination nifi-registry.properties file containing protected config values. :: class java.lang.String ]]. It should only be used with -r and one of [-p, -k, -b].
```
 
+1, merging.
